### PR TITLE
Switch ocean internally to use ocean.meta

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**
First step towards eventual deprecation of ocean.core.Traits

### CODE REVIEW

**Comment by mihails-strasuns-sociomantic** [comment]
(and last big change for v4.0.0)

File: src/ocean/core/TypeConvert.d [hunk 1]
```diff
@@ -192,7 +191,7 @@ template arrayOf (T)
     T[] arrayOf (U...) (U original)
     {
         static assert (U.length > 0);
-        static assert (!hasIndirections!(U));
+        static assert (!hasIndirections!(U[0]));
```
Comment (david-eckardt-sociomantic) [inline src/ocean/core/TypeConvert.d:194]: Is this a separate bugfix?

File: src/ocean/text/convert/Formatter.d [hunk 2]
```diff
@@ -532,8 +536,10 @@ private void handle (T) (T v, FormatInfo f, FormatterSink sf, ElemSink se)
     }
 
     // Arrays (dynamic and static)
-    else static if (is (T A : A[]))
+    else static if (isBasicArrayType!(T))
```
Comment (david-eckardt-sociomantic) [inline src/ocean/text/convert/Formatter.d:539]: Does
```d
 static if (is(ElementTypeOf!(T) A))
```
work? ;)

File: src/ocean/util/DeepReset.d [hunk 3]
```diff
@@ -186,18 +190,18 @@ public void StructDeepReset ( T ) ( ref T dst )
...
-        else static if ( is(typeof(member) S : S[]) )
+        else static if ( isArrayType!(typeof(member)) )
         {
-            static if ( is(typeof(member) U == S[]) ) // dynamic array
+            static if ( isArrayType!(typeof(member)) == ArrayKind.Dynamic )
             {
                 DynamicArrayDeepReset(dst.tupleof[i]);
             }
-            else // static array
+            else
```
Comment (david-eckardt-sociomantic) [inline src/ocean/util/DeepReset.d]: Maybe refactor this to
```d
else static if ( isArrayType!(typeof(member)) == ArrayKind.Dynamic )
    DynamicArrayDeepReset(dst.tupleof[i]);
else static if ( isArrayType!(typeof(member)) == ArrayKind.Static )
    StaticArrayDeepReset(dst.tupleof[i]);
```

File: src/ocean/util/config/ConfigFiller.d [hunk 4]
```diff
@@ -671,38 +669,31 @@ struct SetInfo ( T )
...
+private bool IsSupportedImpl ( T ) ( )
 {
     static if ( is(T : bool) )
-    {
-        const IsSupported = true;
-    }
+        return true;
     else static if ( isIntegerType!(T) || isRealType!(T) )
+        return true;
+    else static if ( isArrayType!(T) )
```
Comment (david-eckardt-sociomantic) [inline src/ocean/util/config/ConfigFiller.d]: If `static if (is(ElementTypeOf!(T) U))` works then you don't need to add the `IsSupportedImpl` function.

File: src/ocean/util/config/ConfigFiller.d [hunk 5]
```diff
@@ -1047,7 +1038,7 @@ private void readFieldsImpl ( T, Source )
 
         if ( config.exists(group, key) )
         {
-            static if ( is(Type U : U[]) && !isStringType!(Type))
+            static if (is(Type U : U[]) && !isUTF8StringType!(Type))
```
Comment (david-eckardt-sociomantic) [inline src/ocean/util/config/ConfigFiller.d:1034]: I'm surprised you keep using `is(Type U : U[])` here.

File: src/ocean/text/convert/Formatter.d (hunk 2 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/text/convert/Formatter.d:539]: Yep (great idea btw!), but it is not very suitable here as `ElementTypeOf` also works on associative arrays.

File: src/ocean/util/config/ConfigFiller.d (hunk 5 above)
Comment (mihails-strasuns-sociomantic) [reply src/ocean/util/config/ConfigFiller.d:1034]: My primary goal was right now is to stop using `core.Traits`, not improve generic facilities in general (that is a bit more complicated task), so I haven't been paying much attention to cases like this.

**Comment by mihails-strasuns-sociomantic** [comment]
Rebased, updated per @david-eckardt-sociomantic comments, fixed D2 build.

File: src/ocean/util/config/ConfigFiller.d (hunk 5 above)
Comment (david-eckardt-sociomantic) [reply src/ocean/util/config/ConfigFiller.d:1034]: Fair enough.

**Review by david-eckardt-sociomantic — APPROVED** [review]
Ick kann nich meckern.

**Pull request merged**